### DISCUSSION TITLE
API with Prisma runs slow on Vercel

### DISCUSSION THREAD

**Discussion by LucasWinkler** [vercel/community#1880] [Help] [answered]
### Summary

I have a Next.js 13 API route that needs to be hit at 12 am EST, once per day only (I hit this API via easycron). It seems that it cold starts after 5 minutes of not being used correct? Well I hit the API and it usually times out due to hitting the hobby limit of 10 seconds for serverless functions. So I attempt to warmup the API route a minute before with another cron job that returns early. I just pass a token and if that token was passed to the route then I return a status 200 and put a message saying warmed up. This actually isn't warming it up because I then hit the API with the correct token for running the logic and it times out or takes extremely long again (7-10+). After actually running the logic and then running the API again it takes anywhere from 4.5-7 seconds. It's warmed up so its no longer timing out but it's still taking longer than it should and it only warms up if I run the whole function.

I don't really understand what's happening. When I run my web app locally and use postman to hit the API it takes just under 2 seconds as seen here:
![image](https://user-images.githubusercontent.com/43054519/227804151-0f3b9a3c-2e05-434c-b563-6e33426dde2e.png)

I'm doing a findMany on a table that will have have up to maybe 8 rows eventually, and currently has 5, I'm also including a one to many relationship but only taking one from it (which for testing purposes only has 2 rows), I then do an upsert and an update. There's no way this should be taking 10 seconds and as seen by this picture it doesn't.

### Additional information

Next.js 13.1.6
Prisma 4.10.1 (Swapped to prismas data proxy and it hasn't helped)
Planetscale (MySQL)

### Example

**Edit:** 
Here is the API route: https://github.com/LucasWinkler/rankguess/blob/main/src/pages/api/select-game-clips-and-revalidate.ts
Gonna say this again but I'm newer to Next.js so if you see anything awful here let me know please! The games query that is within the if statement at the start of the handler is just to test this issue. It will be removed.

**Comment by leerob** [comment]
What region is your Vercel Function and your database in?

**Reply by LucasWinkler** [reply]
Seems to be these

Database: AWS us-east-1 (Northern Virginia)
Vercel: Washington, D.C., USA (iad1)

**Reply by leerob** [reply]
What is the performance if you just return hello world from the function?

**Reply by LucasWinkler** [reply]
Checked if a token from my env variables was the one passed to the API and returned just "Warmed up" and it's extremely quick.

I just did the same thing but included this query to test:
```typescript
const games = await prisma.game.findMany({
      where: {
        isEnabled: true,
      },
      include: {
        clips: {
          where: {
            isAccepted: true,
            hasBeenFeatured: false,
          },
          orderBy: {
            acceptedDate: 'asc',
          },
          take: 1,
        },
        currentClip: true,
      },
    });
```

It took 3.6s-4s. (edit: did another cold start with just that query and it took around 2 seconds. results differ every time) I don't know if it's my findMany, planetscale or prisma. I currently only have 5 rows of Games and there will most likely only ever be a max of 8. There's only 2 clip rows in the database as well. I also just recently switched to using prismas data proxy but that hasn't seemed to help.

**Edit:**
Is it possible that the sorting of clips through the query takes a lot of work even with barely any rows? Would it maybe be better to do it in typescript after just getting the enabled games?

I'm new to this but it's weird that when I run the web app locally, but still connected to the same cloud database it only takes about 2 seconds to run all the logic.

**Reply by leerob** [reply]
Got it, it sounds like this is related to your Prisma query then. Your function and database are in the right region ✅ and the hello world example of your function is fast ✅ so now working backward, there's something with your query here that's slowing things down.

**Reply by LucasWinkler** [reply]
I'll try to mess around with it later but would you have any idea why it might be slow? Does it look off in any way to you? When I get to it I'll possibly try just pulling all the games where they're enabled and then doing the sorting after. Again these literally 0-5 rows currently in any of my tables so it's odd that they would be so slow at times but the cold start definitely impacts the time it takes to run the function by a lot.

Edit: Also the queries run incredibly fast when I use postman on the site that's running locally. The whole function only takes about 2 seconds to run. The database is the same one I'm using on prod so it's not fast because of a local db

**Reply by leerob** [reply]
Another thing to try would be using PlanetScale directly through their serverless driver. You could also try adding indexes to the columns used in the query, but not sure if that's needed due to the dataset size here.

**Reply by LucasWinkler** [reply]
Yeah I'll try their serverless driver. I believe I need to use actual sql queries instead of prisma when using it but I'll try it out for this api route sometime.

**Reply by LucasWinkler** [reply]
Okay so I rewrote my API and I believe it's faster now. The numbers are looking at around 4.5 seconds for a cold start and 2.5-3 seconds after. I'll sleep on it and check again when I wakeup to make sure its not a fluke.

**Comment by LucasWinkler** [comment, marked as answer]
Completely rewrote the API, removed the whole on-demand ISR part of the logic. Decided to just render the pages that used that API as SSR. The API route now takes only a couple seconds after being warmed up. The issue really was just me calling res.revalidate in the API. I'm sure I could use some sort of queue to revalidate after this API is hit but I'll just switch to SSR and everything works perfect. I originally wanted to just use ISR/on-demand ISR because the page/data only updates once per day but in the end I'm happy with it.

Appreciate the replies @leerob

**Reply by leerob** [reply]
Glad to hear it's working for ya!

**Reply by monika-12344** [reply]
Hello actually I am using using one api when button is click and i made the api in next js and also calling through in nextjs as api is in api folder of next js and i am using prisma as planet scale database but when i test that api it runs in 2 to 3sec but when i run into vercel it takes more than 10 sec and fully api response not came can you please help me to resolve this issue?

**Comment by monika-12344** [comment]
Hello actually I am using using one api when button is click and i made the api in next js and also calling through in nextjs as api is in api folder of next js and i am using prisma as planet scale database but when i test that api it runs in 2 to 3sec in postman but when i run into vercel it takes more than 10 sec and fully api response not came can you please help me to resolve this issue?
